fix: Remove bufsize argument from subprocess.Popen

matthewfeickert opened the pull request:
Resolves #77 

Running

```console
pydoc subprocess.Popen
```

shows

```
subprocess.Popen(args, bufsize=-1, ...
```

As "line buffering (buffering=1) isn't supported in binary mode", then remove the line buffering by removing the `bufsize` and letting it default to `-1`.

c.f. https://docs.python.org/3/library/subprocess.html#subprocess.Popen especially

> bufsize will be supplied as the corresponding argument to the [`open()`](https://docs.python.org/3/library/functions.html#open) function when creating the stdin/stdout/stderr pipe file objects:
> 
> * 0 means unbuffered (read and write are one system call and can return short)
> * 1 means line buffered (only usable if `universal_newlines=True` i.e., in a text mode)
> * any other positive value means use a buffer of approximately that size
> * negative bufsize (the default) means the system default of io.DEFAULT_BUFFER_SIZE will be used.

@lukasheinrich The use of `bufsize=1` was added all the way back in 2017 in  https://github.com/yadage/packtivity/commit/3c309d44e76d92714989f810dbf2a5277950348d, so you might have a very good reason for needing that. If you can elaborate that would be great.

matthewfeickert commented:
@lukasheinrich This is ready for review.

lukasheinrich commented:
Hm - why is this  in binary mode anyways? This should be writing out logs so I'm confused why it's not strings

matthewfeickert commented:
> Hm - why is this in binary mode anyways? This should be writing out logs so I'm confused why it's not strings

This is for the `subprocess.Popen` call, which [the docs](https://docs.python.org/3/library/subprocess.html#subprocess.Popen) explain as

> bufsize will be supplied as the corresponding argument to the [`open()`](https://docs.python.org/3/library/functions.html#open) function when creating the stdin/stdout/stderr pipe file objects:
> 
> * 0 means unbuffered (read and write are one system call and can return short)
> * 1 means line buffered (only usable if `universal_newlines=True` i.e., in a text mode)
> * any other positive value means use a buffer of approximately that size
> * negative bufsize (the default) means the system default of `io.DEFAULT_BUFFER_SIZE` will be used.

Our inputs are `argv` which the docs mention as

> args should be a sequence of program arguments or else a single string or path-like object.

and for us are a list of `str`

```diff
$ git diff origin/master 
diff --git a/packtivity/handlers/execution_handlers.py b/packtivity/handlers/execution_handlers.py
index 678e27a..753030e 100644
--- a/packtivity/handlers/execution_handlers.py
+++ b/packtivity/handlers/execution_handlers.py
@@ -297,12 +297,12 @@ def execute_and_tail_subprocess(
                 log.debug("stdin: \n%s", stdin_content)
                 argv = shlex.split(command_string)
                 log.debug("argv: %s", argv)
+                print(f"{argv} is of type {type(argv)}")
                 proc = subprocess.Popen(
                     argv,
                     stdin=subprocess.PIPE,
                     stderr=subprocess.STDOUT,
                     stdout=subprocess.PIPE,
-                    bufsize=1,
                     close_fds=True,
                 )
                 proc.stdin.write(stdin_content.encode("utf-8"))
@@ -312,7 +312,6 @@ def execute_and_tail_subprocess(
                     shlex.split(command_string),
                     stderr=subprocess.STDOUT,
                     stdout=subprocess.PIPE,
-                    bufsize=1,
                     close_fds=True,
                 )
 
```

```console
['docker', 'run', '--rm', '-i', '--cidfile', '/tmp/helloworld/recast-debug/hello_world/_packtivity/hello_world.cid', '-v', '/tmp/helloworld/recast-debug/hello_world:/tmp/helloworld/recast-debug/hello_world:rw', '-v', '/tmp/helloworld/recast-debug/init:/tmp/helloworld/recast-debug/init:rw', 'busybox:1.33', 'sh', '-c', 'sh'] is of type <class 'list'>
```

matthewfeickert commented:
@lukasheinrich Along with what I wrote above, if 

```diff
$ git diff origin/master 
diff --git a/packtivity/handlers/execution_handlers.py b/packtivity/handlers/execution_handlers.py
index 678e27a..c065b27 100644
--- a/packtivity/handlers/execution_handlers.py
+++ b/packtivity/handlers/execution_handlers.py
@@ -303,9 +303,10 @@ def execute_and_tail_subprocess(
                     stderr=subprocess.STDOUT,
                     stdout=subprocess.PIPE,
                     bufsize=1,
+                    universal_newlines=True,
                     close_fds=True,
                 )
-                proc.stdin.write(stdin_content.encode("utf-8"))
+                proc.stdin.write(stdin_content)
                 proc.stdin.close()
             else:
                 proc = subprocess.Popen(
@@ -313,6 +314,7 @@ def execute_and_tail_subprocess(
                     stderr=subprocess.STDOUT,
                     stdout=subprocess.PIPE,
                     bufsize=1,
+                    universal_newlines=True,
                     close_fds=True,
                 )
 
```

was used instead it gets stuck in a loop where it should have finished in seconds

```
2021-12-15 10:16:03,655 | packtivity.asyncback |   INFO | configured pool size to 12
2021-12-15 10:16:03,687 |      yadage.creators |   INFO | initializing workflow with initdata: {'name': 'hello'} discover: True relative: True
2021-12-15 10:16:03,687 |    adage.pollingexec |   INFO | preparing adage coroutine.
2021-12-15 10:16:03,687 |                adage |   INFO | starting state loop.
2021-12-15 10:16:03,738 |     yadage.wflowview |   INFO | added </init:0|defined|unknown>
2021-12-15 10:16:04,401 |     yadage.wflowview |   INFO | added </hello_world:0|defined|unknown>
2021-12-15 10:16:05,260 |    adage.pollingexec |   INFO | submitting nodes [</init:0|defined|known>]
2021-12-15 10:16:05,720 |       pack.init.step |   INFO | publishing data: <TypedLeafs: {'name': 'hello'}>
2021-12-15 10:16:05,720 |                adage |   INFO | unsubmittable: 0 | submitted: 0 | successful: 0 | failed: 0 | total: 2 | open rules: 0 | applied rules: 2
2021-12-15 10:16:06,671 |           adage.node |   INFO | node ready </init:0|success|known>
2021-12-15 10:16:06,671 |    adage.pollingexec |   INFO | submitting nodes [</hello_world:0|defined|known>]
2021-12-15 10:16:06,672 | pack.hello_world.ste |   INFO | starting file logging for topic: step
2021-12-15 10:16:36,259 |                adage |   INFO | unsubmittable: 0 | submitted: 1 | successful: 1 | failed: 0 | total: 2 | open rules: 0 | applied rules: 2
2021-12-15 10:17:06,477 |                adage |   INFO | unsubmittable: 0 | submitted: 1 | successful: 1 | failed: 0 | total: 2 | open rules: 0 | applied rules: 2
2021-12-15 10:17:37,405 |                adage |   INFO | unsubmittable: 0 | submitted: 1 | successful: 1 | failed: 0 | total: 2 | open rules: 0 | applied rules: 2
2021-12-15 10:18:08,334 |                adage |   INFO | unsubmittable: 0 | submitted: 1 | successful: 1 | failed: 0 | total: 2 | open rules: 0 | applied rules: 2
2021-12-15 10:18:39,394 |                adage |   INFO | unsubmittable: 0 | submitted: 1 | successful: 1 | failed: 0 | total: 2 | open rules: 0 | applied rules: 2
...
^C
```

matthewfeickert commented:
Approved over Skype so merging.
